File: src/main/java/com/rephilo/luandun/leetcode/_278/Solution.java
package com.rephilo.luandun.leetcode._278;

public class Solution {
    /**
     * 278
     *
     * @param n
     * @return
     */
    public int firstBadVersion(int n) {
        int left = 0;
        int right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    private boolean isBadVersion(int n) {
        return false;
    }
}

File: src/main/scala/com/rephilo/luandun/scala/leetcode/_278/Solution.scala
package com.rephilo.luandun.scala.leetcode._278

object Solution {
  /**
    * 278
    *
    * @param n
    * @return
    */
  def firstBadVersion(n: Int): Int = {
    val left = 1
    val right = n

    @scala.annotation.tailrec
    def calc(left: Int, right: Int): Int = {
      if (left >= right) {
        return left
      } else {
        val mid = left + (right - left) / 2
        if (isBadVersion(mid)) {
          calc(left, mid)
        } else {
          calc(mid + 1, right)
        }
      }
    }

    def isBadVersion(n: Int): Boolean = {
      return false
    }

    calc(left, right)
  }
}
